File: protocol-generator/php/src/Conformance.php
<?php

declare(strict_types=1);

namespace EntityCore;

/**
 * ECF conformance runner. Pure (no file IO): it takes the corpus bytes, decodes
 * them with THIS peer's own decoder (a decoder bug here is itself a conformance
 * failure, ENTITY-CBOR-ENCODING.md §E.3), and runs every encode_equal /
 * decode_reject vector.
 *
 * Vectors dispatch by category (the id prefix before the dot):
 *   * content_hash — varint(format_code) || SHA-2(ECF({type, data}))
 *   * peer_id      — Cbor.encode(Base58(varint(kt)||varint(ht)||digest)) as text
 *   * signature    — Ed25519_sign(seed, ECF(entity))
 *   * everything else (float/int/map_keys/length/primitive/nested/envelope)
 *     — plain Cbor.encode(input)
 *   * decode_reject — the decoder MUST reject the canonical wire bytes
 */
final class Conformance
{
    private const KINDS = ['encode_equal', 'decode_reject'];

    /**
     * Run every encode_equal / decode_reject vector in the decoded corpus.
     *
     * @return list<ConformanceResult>
     */
    public static function run(string $corpusBytes): array
    {
        $vectors = Cbor::decode($corpusBytes);
        if (!\is_array($vectors)) {
            throw new CodecException('corpus did not decode to an array of vectors');
        }
        $results = [];
        foreach ($vectors as $v) {
            if (!$v instanceof EcfMap) {
                continue; // meta rows that are not maps
            }
            $kind = $v->get('kind');
            if (!\is_string($kind) || !\in_array($kind, self::KINDS, true)) {
                continue; // meta / non-conformance rows
            }
            $results[] = self::runVector($v);
        }
        return $results;
    }

    private static function runVector(EcfMap $vector): ConformanceResult
    {
        $id = (string) $vector->get('id');
        $kind = (string) $vector->get('kind');
        return match ($kind) {
            'decode_reject' => self::runReject($id, $vector->get('canonical')),
            'encode_equal' => self::runEncode($id, $vector->get('input'), $vector->get('canonical')),
            default => new ConformanceResult($id, false, "unknown kind: {$kind}"),
        };
    }

    private static function runReject(string $id, mixed $wire): ConformanceResult
    {
        $bytes = self::asBytes($wire);
        try {
            Cbor::decode($bytes);
            return new ConformanceResult($id, false, 'expected reject but decoded successfully');
        } catch (CodecException) {
            return new ConformanceResult($id, true, null);
        }
    }

    private static function runEncode(string $id, mixed $input, mixed $want): ConformanceResult
    {
        $wantBytes = self::asBytes($want);
        try {
            $got = self::produce($id, $input);
        } catch (\Throwable $e) {
            return new ConformanceResult($id, false, 'raised ' . $e::class . ': ' . $e->getMessage());
        }
        if ($got === $wantBytes) {
            return new ConformanceResult($id, true, null);
        }
        return new ConformanceResult(
            $id,
            false,
            'got ' . \bin2hex($got) . ' want ' . \bin2hex($wantBytes)
        );
    }

    private static function produce(string $id, mixed $input): string
    {
        $category = \explode('.', $id)[0];
        return match ($category) {
            'content_hash' => self::produceContentHash($input),
            'peer_id' => self::producePeerId($input),
            'signature' => self::produceSignature($input),
            default => Cbor::encode($input),
        };
    }

    private static function produceContentHash(mixed $input): string
    {
        if (!$input instanceof EcfMap) {
            throw new UnsupportedValueException('content_hash input must be a map');
        }
        $entity = ['type' => $input->get('type'), 'data' => $input->get('data')];
        $formatCode = 0;
        if ($input->hasTextKey('format_code')) {
            $fc = $input->get('format_code');
            $formatCode = \is_int($fc) ? $fc : (int) \gmp_strval($fc);
        }
        return Hash::contentHash($entity, $formatCode);
    }

    private static function producePeerId(mixed $input): string
    {
        if (!$input instanceof EcfMap) {
            throw new UnsupportedValueException('peer_id input must be a map');
        }
        $keyType = self::asInt($input->get('key_type'));
        $hashType = self::asInt($input->get('hash_type'));
        $digest = self::asBytes($input->get('digest'));
        $peer = PeerId::format($keyType, $hashType, $digest);
        // The corpus pins the Base58 peer-id string ECF-encoded as CBOR text.
        return Cbor::encode($peer);
    }

    private static function produceSignature(mixed $input): string
    {
        if (!$input instanceof EcfMap) {
            throw new UnsupportedValueException('signature input must be a map');
        }
        $seed = self::asBytes($input->get('seed'));
        $entity = $input->get('entity');
        return Signature::signRaw($seed, Cbor::encode($entity));
    }

    /** Coerce a decoded value to raw bytes (ByteString or already a string). */
    private static function asBytes(mixed $v): string
    {
        if ($v instanceof ByteString) {
            return $v->bytes;
        }
        if (\is_string($v)) {
            return $v;
        }
        throw new UnsupportedValueException('expected bytes, got ' . \get_debug_type($v));
    }

    private static function asInt(mixed $v): int
    {
        if (\is_int($v)) {
            return $v;
        }
        if ($v instanceof \GMP) {
            return (int) \gmp_strval($v);
        }
        throw new UnsupportedValueException('expected int, got ' . \get_debug_type($v));
    }
}

File: protocol-generator/zig/src/conformance.zig
//! ECF conformance harness — loads the normative fixture
//! (conformance-vectors-v1.cbor) and runs every vector through the codec,
//! checking byte-identity (encode_equal) or rejection (decode_reject) per
//! Appendix E §E.3. The fixture carries its own cross-blessed `canonical` bytes
//! (produced + 3-way cross-blessed by the Go/Rust/Python oracles), so this is
//! self-contained — no running Go oracle needed at S2 (the Go wire-conformance
//! binary is the fixture producer, not a runtime checker). std.testing.allocator
//! is NOT used here (this is a CLI exe, not a test); a GeneralPurposeAllocator
//! with leak detection backs it instead so any leak is reported on exit.

const std = @import("std");
const codec = @import("root.zig");
const cbor = codec.cbor;

const Value = cbor.Value;

fn mapGet(m: []Value.Pair, name: []const u8) ?Value {
    for (m) |p| {
        switch (p.key) {
            .text => |t| if (std.mem.eql(u8, t, name)) return p.value,
            else => {},
        }
    }
    return null;
}

fn asMap(v: Value) []Value.Pair {
    return switch (v) {
        .map => |m| m,
        else => unreachable,
    };
}
fn asText(v: Value) []const u8 {
    return switch (v) {
        .text => |t| t,
        else => unreachable,
    };
}
fn asBytes(v: Value) []const u8 {
    return switch (v) {
        .bytes => |b| b,
        else => unreachable,
    };
}
fn asUint(v: Value) u64 {
    return switch (v) {
        .uint => |n| n,
        else => unreachable,
    };
}

fn category(id: []const u8) []const u8 {
    if (std.mem.indexOfScalar(u8, id, '.')) |i| return id[0..i];
    return id;
}

const Outcome = union(enum) { pass, fail: []const u8 };

/// Run one vector. `scratch` is a per-vector arena (freed by the caller); any
/// produced bytes live in it, so we never leak across vectors.
fn runVector(scratch: std.mem.Allocator, vm: []Value.Pair) !Outcome {
    const id = asText(mapGet(vm, "id").?);
    const kind = asText(mapGet(vm, "kind").?);
    const canon = asBytes(mapGet(vm, "canonical").?);
    const cat = category(id);

    if (std.mem.eql(u8, kind, "decode_reject")) {
        // The decoder MUST reject these wire bytes (tags / non-canonical).
        if (cbor.decode(scratch, canon)) |v| {
            v.deinit(scratch);
            return .{ .fail = "decoder accepted a reject vector" };
        } else |_| {
            return .pass;
        }
    }

    // encode_equal: produce bytes per category, compare to canon.
    var produced: []u8 = undefined;
    if (std.mem.eql(u8, cat, "content_hash")) {
        const input = asMap(mapGet(vm, "input").?);
        const typ = asText(mapGet(input, "type").?);
        const data = mapGet(input, "data").?;
        const fc: u64 = if (mapGet(input, "format_code")) |v| asUint(v) else 0;
        produced = try codec.hash.contentHash(scratch, fc, typ, data);
    } else if (std.mem.eql(u8, cat, "peer_id")) {
        const input = asMap(mapGet(vm, "input").?);
        const key_type = asUint(mapGet(input, "key_type").?);
        const hash_type = asUint(mapGet(input, "hash_type").?);
        const digest = asBytes(mapGet(input, "digest").?);
        const pid = try codec.peer_id.format(scratch, .{ .key_type = key_type, .hash_type = hash_type, .digest = digest });
        // canonical bytes are the ECF encoding of the peer-id text string.
        produced = try cbor.encode(scratch, .{ .text = pid });
    } else if (std.mem.eql(u8, cat, "signature")) {
        const input = asMap(mapGet(vm, "input").?);
        const seed_b = asBytes(mapGet(input, "seed").?);
        const entity = asMap(mapGet(input, "entity").?);
        const typ = asText(mapGet(entity, "type").?);
        const data = mapGet(entity, "data").?;
        const msg = try codec.hash.ecfOfEntity(scratch, typ, data);
        var seed: [32]u8 = undefined;
        @memcpy(&seed, seed_b[0..32]);
        const sig = try codec.sign.sign(seed, msg);
        produced = try scratch.dupe(u8, &sig);
    } else {
        // float / int / map_keys / length / primitive / nested / envelope:
        // re-encode the decoded input value canonically.
        produced = try cbor.encode(scratch, mapGet(vm, "input").?);
    }

    if (std.mem.eql(u8, produced, canon)) return .pass;
    const msg = try std.fmt.allocPrint(scratch, "want {x} got {x}", .{ canon, produced });
    return .{ .fail = msg };
}

pub fn main() !void {
    var gpa_state: std.heap.GeneralPurposeAllocator(.{}) = .init;
    defer _ = gpa_state.deinit();
    const gpa = gpa_state.allocator();

    var args = try std.process.argsWithAllocator(gpa);
    defer args.deinit();
    _ = args.next(); // exe name
    const path = args.next() orelse "../shared/test-vectors/v0.8.0/conformance-vectors-v1.cbor";

    const file = try std.fs.cwd().openFile(path, .{});
    defer file.close();
    const bytes = try file.readToEndAlloc(gpa, 1 << 24);
    defer gpa.free(bytes);

    const fixture = try cbor.decode(gpa, bytes);
    defer fixture.deinit(gpa);
    const vectors = switch (fixture) {
        .array => |a| a,
        else => return error.FixtureNotArray,
    };

    const stdout = std.fs.File.stdout();
    var pass: usize = 0;
    var fail: usize = 0;

    // per-category tallies
    var cats: std.StringArrayHashMapUnmanaged([2]usize) = .empty;
    defer cats.deinit(gpa);

    for (vectors) |vraw| {
        const vm = asMap(vraw);
        const id = asText(mapGet(vm, "id").?);
        const cat = category(id);

        // per-vector arena — everything runVector allocates dies here.
        var arena = std.heap.ArenaAllocator.init(gpa);
        const outcome = try runVector(arena.allocator(), vm);

        const gop = try cats.getOrPut(gpa, cat);
        if (!gop.found_existing) gop.value_ptr.* = .{ 0, 0 };
        switch (outcome) {
            .pass => {
                pass += 1;
                gop.value_ptr.*[0] += 1;
            },
            .fail => |m| {
                fail += 1;
                gop.value_ptr.*[1] += 1;
                var buf: [512]u8 = undefined;
                const line = std.fmt.bufPrint(&buf, "FAIL {s}  {s}\n", .{ id, m }) catch "FAIL (msg too long)\n";
                _ = stdout.writeAll(line) catch {};
            },
        }
        arena.deinit();
    }

    var buf: [256]u8 = undefined;
    _ = stdout.writeAll("\n-- by category --\n") catch {};
    var it = cats.iterator();
    while (it.next()) |e| {
        const p = e.value_ptr.*[0];
        const f = e.value_ptr.*[1];
        const line = std.fmt.bufPrint(&buf, "  {s:<14} {d}/{d}\n", .{ e.key_ptr.*, p, p + f }) catch continue;
        _ = stdout.writeAll(line) catch {};
    }
    const total = std.fmt.bufPrint(&buf, "\nTOTAL: {d} passed, {d} failed (of {d})\n", .{ pass, fail, pass + fail }) catch "";
    _ = stdout.writeAll(total) catch {};

    if (fail > 0) std.process.exit(1);
}
